scala/lambdas/postingest-state-change-handler/src/main/scala/uk/gov/nationalarchives/postingeststatechangehandler/Lambda.scala

```diff
@@ -0,0 +1,213 @@
+    def getPrimaryKey(item: PostIngestStatusTableItem) =
+      FilesTablePrimaryKey(FilesTablePartitionKey(item.assetId), FilesTableSortKey(item.batchId))
+
+    def addOrUpdateItem(item: PostIngestStatusTableItem, queueAlias: String): IO[Unit] = {
+      val batchId = AttributeValue.builder().s(item.batchId).build()
+      val postIngestQueue = AttributeValue.builder().s(queueAlias).build()
+      val dateTimeNow = dependencies.instantGenerator()
+      val dateTimeNowIso = AttributeValue.builder().s(dateTimeNow.toString).build()
+
+      dependencies.daDynamoDbClient
+        .updateAttributeValues(
+          DADynamoDbRequest(
+            config.dynamoTableName,
+            Map("assetId" -> AttributeValue.builder().s(item.assetId.toString).build()),
+            Map("queue" -> Some(postIngestQueue), "firstQueuedAt" -> Some(dateTimeNowIso), "lastQueuedAt" -> Some(dateTimeNowIso))
+          )
+        )
+        .void
+    }
+
+    def deleteItemFromTable(item: PostIngestStatusTableItem) =
+      dependencies.daDynamoDbClient.deleteItems(config.dynamoTableName, List(getPrimaryKey(item))).void
+
+    def sendMessageToQueue(queueUrl: String, message: OutputQueueMessage): IO[Unit] =
+      dependencies.daSqsClient.sendMessage(queueUrl)(message).void
+
+    def sendOutputMessage(item: PostIngestStatusTableItem, newQueueAlias: Option[String] = None): IO[Unit] = {
+      val (messageType, messageStatus) = newQueueAlias match
+        case Some("CC")             => (IngestUpdate, IngestedPreservation)
+        case Some(unsupportedQueue) => throw new Exception(s"A 'messageType' and 'messageStatus' implementation exist for queue $unsupportedQueue")
+        case None                   => (IngestComplete, IngestedCCDisk)
+
+      val message = OutputMessage(
+        OutputProperties(item.batchId, dependencies.uuidGenerator(), item.correlationId, dependencies.instantGenerator(), messageType),
+        OutputParameters(item.assetId, messageStatus)
+      )
+      dependencies.daSnsClient.publish(config.topicArn)(message :: Nil).void
+    }
+
+    def updateTableAndSendToSqs(newItem: PostIngestStatusTableItem, queue: Queue) =
+      addOrUpdateItem(newItem, queue.queueAlias) >>
+        sendMessageToQueue(
+          queue.queueUrl,
+          OutputQueueMessage(newItem.assetId, newItem.batchId, queue.resultAttrName, newItem.input)
+        )
+
+    def getInsertFibers(queues: List[Queue]) = event.Records
+      .filter(_.eventName == EventName.INSERT)
+      .parTraverse { record =>
+        val queue1 = queues.find(_.queueOrder == 1).get
+        val processInsertRecord = updateTableAndSendToSqs(record.dynamodb.newImage, queue1) >> sendOutputMessage(record.dynamodb.newImage, Some(queue1.queueAlias))
+        processInsertRecord.start
+      }
+
+    def newResultDiffersFromOld(newResult: Option[String], oldResult: Option[String]) = newResult.getOrElse("") != oldResult.getOrElse("")
```

Comment:
You don't need to do getOrElse, if you compare `Option` directly then it works and if you do that you don't really need a seperate method either.

Reply:
Wait! Both of them can be `None` if for example a different attribute gets changed e.g.

```
// oldImage is this
PostIngestStatusTableItem(1, "batchId", "input", Some("correlationId"), Some(queue1), "dateTimeYesterday", "dateTimeYesterday", potentialResultCC=None)
```
and 
```

// newImage is this
PostIngestStatusTableItem(1, "batchId", "input", Some("correlationId"), Some(queue1), "dateTimeToday", "dateTimeToday", potentialResultCC=None)
```

This would trigger a MODIFY event, right?

And then when you do a `!newResult.contains(oldResult)` (i.e `!newPotentialResultCC.contains(oldPotentialResultCC)`), you would get `true`